### PR TITLE
Nicer format for whoami and sorting on groups.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/25/
Test FAILed.

**Comment by fabricematrat** [comment]
QA:
https://github.com/juju/charmstore-client/issues/22
https://github.com/juju/charmstore-client/issues/15

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/26/
Test PASSed.

File: cmd/charm/charmcmd/whoami.go [hunk 1]
```diff
@@ -8,12 +8,18 @@ import (
 	"net/url"
 	"strings"
 
+	"bytes"
```
Comment (frankban) [inline cmd/charm/charmcmd/whoami.go]: Please move this to the first import group.

File: cmd/charm/charmcmd/whoami.go [hunk 2]
```diff
@@ -8,12 +8,18 @@ import (
...
 	"github.com/juju/cmd"
 	"gopkg.in/errgo.v1"
+	"gopkg.in/juju/charmrepo.v2-unstable/csclient/params"
+	"launchpad.net/gnuflag"
+	"sort"
```
Comment (frankban) [inline cmd/charm/charmcmd/whoami.go]: This one too.

File: cmd/charm/charmcmd/whoami.go (hunk 1 above)
Comment (mhilton) [inline cmd/charm/charmcmd/whoami.go]: this one goes in the first section

File: cmd/charm/charmcmd/whoami.go [hunk 3]
```diff
@@ -8,12 +8,18 @@ import (
...
+	"text/tabwriter"
```
Comment (frankban) [inline cmd/charm/charmcmd/whoami.go]: Ditto.

File: cmd/charm/charmcmd/whoami.go (hunk 2 above)
Comment (mhilton) [inline cmd/charm/charmcmd/whoami.go]: as do these two

File: cmd/charm/charmcmd/whoami_test.go [hunk 4]
```diff
@@ -85,7 +85,7 @@ func (s *whoamiSuite) TestLoggedIn(c *gc.C) {
 	stdout, stderr, exitCode := run(c.MkDir(), "whoami")
 	c.Assert(stderr, gc.Equals, "")
 	c.Assert(exitCode, gc.Equals, 0)
-	c.Assert(stdout, gc.Equals, "test-user\ntest-group1 test-group2\n")
+	c.Assert(stdout, gc.Equals, "User: test-user\nGroup membership: test-group1, test-group2\n")
```
Comment (frankban) [inline cmd/charm/charmcmd/whoami_test.go:89]: Can we make it clear that groups are ordered alphabetically in the test?
Could you also add tests for other --format options?

**Comment by frankban** [comment]
:+1: with some minors and little more testing.
Thank you!

File: cmd/charm/charmcmd/whoami.go [hunk 5]
```diff
@@ -31,6 +37,27 @@ func (c *whoamiCommand) Info() *cmd.Info {
...
+		"yaml": cmd.FormatYaml,
+		"json": cmd.FormatJson,
+		"text": c.formatText,
+	})
+}
+
+func (c *whoamiCommand) formatText(value interface{}) ([]byte, error) {
+	resp := value.(*params.WhoAmIResponse)
+	var out bytes.Buffer
+	tw := tabwriter.NewWriter(&out, 0, 1, 1, ' ', 0)
```
Comment (rogpeppe) [inline cmd/charm/charmcmd/whoami.go]: Why use tabwriter if we're not emitting any tabs?

**Comment by rogpeppe** [comment]
Looks reasonable, but I have my doubts about showing groups. Specifically, putting groups in the output of this command blesses the idea that one can enumerate group membership. That's something that is trivially done now, but it's architecturally problematic, constraining possible implementations of the identity manager and group semantics. I'd prefer to add that feature only if there's particular demand, and even then it might be in a different form (for example: querying idm directly).

If we drop group membership from the output, then this command could be much simpler and just print the name of the user to stdout, with no need for different formats.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/35/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/charmstore-client/39/
Test PASSed.

File: cmd/charm/charmcmd/whoami.go (hunk 1 above)
Comment (fabricematrat) [inline cmd/charm/charmcmd/whoami.go]: Done.

File: cmd/charm/charmcmd/whoami.go (hunk 2 above)
Comment (fabricematrat) [inline cmd/charm/charmcmd/whoami.go]: Done.

File: cmd/charm/charmcmd/whoami_test.go (hunk 4 above)
Comment (fabricematrat) [inline cmd/charm/charmcmd/whoami_test.go:89]: Done. Added a test just for sort.

File: cmd/charm/charmcmd/whoami.go (hunk 5 above)
Comment (fabricematrat) [inline cmd/charm/charmcmd/whoami.go]: Removed.

**Comment by urosj** [comment]
:+1: groups, no matter where they are, will still be needed to handle existing cases like ~charmers, etc.

:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-client-merge

**Pull request merged**